feat: export and live update data to Google Sheets

chazkiker2 opened the pull request:
# Description

- [x] close #5
- [x] implements `tokio::sync::mpsc` to separate the concerns of **producing data** vs **consuming data** as suggested by @nikomatsakis 
- [x] users have the ability to attach a google sheets ID by way of the `--google_sheets <ID>` flag or `-g <ID>`. This will convey that they'd like to export data to that Sheet!
- [x] users are re-directed to sign in to Google (via a link printed to terminal) where they can then allow us to take control
- [x] data is exporting to google sheets :smile:
  - [x] when exporting to a sheet, the entirety of the sheet's values (not formatting) will be cleared; then we will insert our data starting with cell `A1`
- [x] a message is conveyed to the user after processing 
- **ARCHIVED** ~~Introduces some first-draft structs, traits, and implementations to manage our various data types and de-couple _gathering_ data vs _outputing_ data.~~
  - ~~I am 100% confident that somewhere in this PR my lack of experience in this language will be exposed. I expect it will greatly benefit from a magical refactor from @nikomatsakis.~~
  - ~~gathering and outputting data for `rust-lang` takes **even longer** now... (or feels like it at least since it currently gathers all data before printing everything — **which we should change eventually**)~~


# Comments 

- the code found in `google_sheets.rs` was initially inspired by the [`sheets` crate](https://crates.io/crates/sheets), because I found their initial design rather simple and clean, albiet a bit lacking in what we needed. 
- Though there is naturally a little overhead by building our google-sheets communications from scratch, I personally think the control is really convenient — especially as we may someday want to provide some Inversion of Control to our users RE: formatting sheets options (just for example), in which case the granular control we have by interacting directly with the Google's API is a pretty powerful and convenient thing.\
Plus, in all honesty, I was not overly impressed with what the two rust crates I looked at had to offer.

# Thoughts / Ideas / Conversations

- the google sheets export depends on an OAuth 2.0 Client ID for Desktop (Server) from a Google Cloud Application. Our `Sheets` client looks for that in a file called `client_secret.json` in the root folder. 
  - This is the Client ID for a GCloud application called `gh-metrics-rst` under ownership of a freshly created GMail account <gh.metrics.rst@gmail.com> that we can use as a default so that users can use the CLI out of the box.
  - Later on, we can support the ability for users to specify their own Client ID or Cloud App
- We should add a "progress bar" so our user doesn't need to wonder what is happening when we're trying to analyze the entirety of `rust-lang`
  - on this note, the code introduced in this PR doesn't log each repo/line one-by-one, therefore you end up waiting around for quite some time just waiting for our CLI to load up all the data.

chazkiker2 commented:
@nikomatsakis I would love your feedback on this. I expect you'll have a decent amount.

Further, I am not overly attached to the data structures introduced here... I kind of felt like a fish out of water without my OOP inheritance, in all honesty. I'm sure it's not the most idiomatic Rust you'll ever see.

Let's keep what you like, improve what you don't, and burn what you hate :smile: !

nikomatsakis commented:
@chazkiker2 great! Sorry I've been slow, I'm at an off-site this week so I'm a bit distracted. I'll try to take a look as soon as I can.

nikomatsakis commented on src/main.rs:
```diff
@@ -20,9 +25,97 @@ enum Opt {
     },
 }
 
+struct Collection<T: google_sheets::IntoSheetEntry> {
```
There's a lot here. At minimum we need another module, but also some comments to explain the intention. I'm going to take a bit of time to read into it but I may hit my quota and have to come back to it.

nikomatsakis commented on src/main.rs:
```diff
@@ -20,9 +25,97 @@ enum Opt {
 }
 
+struct Collection<T: google_sheets::IntoSheetEntry> {
+    headers: Vec<String>,
```
Are these "column names"?

chazkiker2 replied:
Correct.

nikomatsakis commented on src/main.rs:
```diff
@@ -20,9 +25,97 @@ enum Opt {
+struct Collection<T: google_sheets::IntoSheetEntry> {
+    headers: Vec<String>,
+    data: Vec<T>,
+    num_columns: usize,
```
Can't we get this from `headers.len()`?

chazkiker2 replied:
Yes, which we do. This is unused. And we get `num_rows` from `data.len()`

I will remove these.

nikomatsakis commented on src/main.rs:
```diff
@@ -20,9 +25,97 @@ enum Opt {
     },
 }
 
+struct Collection<T: google_sheets::IntoSheetEntry> {
```
One thing I don't love is that `Collection` feels "independent" from google sheets, and yet it mentions google sheets in its where clauses.

chazkiker2 replied:
This is a very good point. Will change.

nikomatsakis commented on src/main.rs:
```diff
@@ -20,9 +25,97 @@ enum Opt {
+    }
+}
+
+impl<T: std::fmt::Display + google_sheets::IntoSheetEntry> Collection<T> {
```
Nit: I prefer

```rust
where
    T: Display + IntoSheetEntry
```

than using this "inline" form.

nikomatsakis reviewed (commented):
OK, I read it and I left a few nits. The google sheets code looks nice, and I wouldn't be opposed to landing this PR, but I think I would want to rework it into more of an "actor" approach, versus building up data structures. This is partly because I want us to be able to print out live data as it is printed, given how slow some things seem to be!

nikomatsakis commented:
I'll try to leave comments on Zulip about the actor approach later on!

chazkiker2 commented on src/google_sheets.rs:
```diff
@@ -0,0 +1,247 @@
+        Sheets::new(token, sheet_id)
+    }
+
+    pub async fn authenticate() -> Result<AccessToken, APIError> {
```
agreed.

chazkiker2 commented on src/google_sheets.rs:
```diff
@@ -0,0 +1,247 @@
+                client,
+                sheet_id: String::from(sheet_id),
+            }),
+            Err(_) => Err(APIError {
```
hmmm yeah okay that makes sense. 

I'll play around with that.

chazkiker2 commented on src/google_sheets.rs:
```diff
@@ -0,0 +1,247 @@
+        }
+    }
+
+    pub fn request<T>(
```
Yes, I should go through and add doc comments everywhere.

This is not exactly for end-users, but I guess it could be? I think I should remove the `pub` for now, though.

chazkiker2 commented on src/google_sheets.rs:
```diff
@@ -0,0 +1,247 @@
+        let url = Url::parse(BASE_ENDPOINT).unwrap().join(&path).unwrap();
+
+        // TODO: use `self.token = Sheets::authenticate().await;` to attempt to update token from cache
+        // NOTE: this would make `request` async and I don't know if we want that or not
```
Went to go change this and realized the reason this was a challenge was less due to the async nature and more due to the assignment to `self.token`.

This would require a mutable reference to `&mut self` for nearly every method in `google_sheets::Sheets`, which would require client code to use `let mut sheets = Sheets::initialize(&sheet_id).await.unwrap()` any time they instantiated it... which maybe that's fine, I'm unsure... but it felt "bad" to me for some reason.

chazkiker2 replied:
we also may not even need this given the way that every sheets instance is set up...

chazkiker2 commented:
@nikomatsakis absolutely no worries on the timeline. Sorry that you're busy, but I hope it's the good kind of busy!!

I implemented almost all of the changes you mentioned.

- [x] my sloppy structures have been completely removed 🥳  (truly something to be thankful for, I think)
- [x] the `tokio::sync::mpsc::channel` is now leading the charge (working very smoothly... that was an awesome recommendation on your part)
- [x] we're now using `snafu` to customize domain-specific errors with context but still preserve source.
- [x] data is outputting progressively as we get it once again — this is true for both the terminal and google sheets ⚡ 
- [x] the `Sheets` implementation now has even more methods (allowing dead code for now, b/c as this grows I think we'll use some of them) 
- [x] I've also added a bit more in the realm of comments; hopefully that is useful.

Thanks for the review. Let me know what you think of this round. (I think it is immensely improved from last time)

nikomatsakis commented on src/main.rs:
```diff
@@ -32,14 +40,95 @@ async fn main() {
+            let (tx, mut rx) = mpsc::channel::<Vec<String>>(400);
+
+            // produce all relevant data within this
+            tokio::spawn(async move {
```
Nit: can we factor this `async move` block into an `async fn`?

nikomatsakis commented on src/main.rs:
```diff
@@ -32,14 +40,95 @@ async fn main() {
+            match gs {
+                // if user specified a google sheet ID, they must want to export data to that sheet!
+                Some(sheet_id) => {
+                    let sheets = match Sheets::initialize(&sheet_id).await {
```
Nit: can we move this match arm to an async fn

nikomatsakis reviewed (changes requested):
This looks great. Left 2 nits.
